target/linux/ath79/image/lzma-loader/src/loader.c

```diff
@@ -243,7 +285,27 @@ void loader_main(unsigned long reg_a0, unsigned long reg_a1,
 		default:
 			printf("unknown error %d!\n", res);
 		}
+#if (LZMA_WRAPPER)
 		halt();
+#else
+		if (CONFIG_FLASH_ALT != 0) {
+			lzma_init_data_alt();
+			res = lzma_init_props();
+			if (res != LZMA_RESULT_OK) {
+				halt();
+			}
```

Comment:
This seems redundant.

Also the code-flow looks not really great. If the first location is corrupt and the second one has an intact LZMA header but corrupt payload, you are effectively trying to boot twice from the same source.

Better: Abstract the boot process to a parameterized function which returns non-zero in case the LZMA decompression fails at some stage. If the loader is expected to try a second location from flash, it can reuse the entire boot process function.

Reply:
thanks a lot for the review and ideas. I'll try some things to make this much cleaner looking